File: etl/utils/scraper.py
from bs4 import BeautifulSoup
import requests
import logging

BASE_URL = 'https://www.boletinoficial.gob.ar'

def today_urls():
    """
    Returns a list of ids that correspond
    to that day's publications.
    """
    articles_url = f'{BASE_URL}/seccion/primera'

    try:
        page = requests.get(articles_url)
        page.raise_for_status()
    except requests.exceptions.RequestException as err:
        logging.error(f"Error occurred: {err}")
        return [], None

    soup = BeautifulSoup(page.content, "html.parser")
    body = soup.find(id='avisosSeccionDiv')

    urls = [
        f'{BASE_URL}{a["href"]}'
        for a in body.find_all("a", href=True)
    ]

    return urls, page.status_code

def scrape_article(article_url):
    """
    Scrapes article from Boletin Oficial.
    Returns Type, Content and Date.
    """
    try:
        page = requests.get(article_url)
        page.raise_for_status()
    except requests.exceptions.RequestException as err:
        logging.error(f"Error occurred: {err}")
        return None, None, None, None

    soup = BeautifulSoup(page.content, "html.parser")

    area = soup.find(id="tituloDetalleAviso").find("h1").text

    article_type = soup.find(class_="puntero first-section").text
    
    content = soup.find(id="cuerpoDetalleAviso").text

    return article_type, area, content, page.status_code

File: internal/scraper/scraper.go
package scraper

import (
	"fmt"
	"net/http"
	"strings"

	"github.com/PuerkitoBio/goquery"
	"github.com/gocolly/colly/v2"
)

func TodayURLs() ([]string, error) {
	var urls []string
	c := colly.NewCollector()

	c.OnHTML("#avisosSeccionDiv a[href]", func(e *colly.HTMLElement) {
		url := "https://www.boletinoficial.gob.ar" + strings.Split(e.Attr("href"), "?")[0]
		urls = append(urls, url)
	})

	err := c.Visit("https://www.boletinoficial.gob.ar/seccion/primera")
	if err != nil {
		return nil, err
	}

	// Remove duplicates
	seen := make(map[string]bool)
	var unique []string
	for _, url := range urls {
		if !seen[url] {
			seen[url] = true
			unique = append(unique, url)
		}
	}
	return unique, nil
}

func ScrapeArticle(url string) (string, string, string, error) {
	resp, err := http.Get(url)
	if err != nil {
		return "", "", "", err
	}
	defer resp.Body.Close()

	doc, err := goquery.NewDocumentFromReader(resp.Body)
	if err != nil {
		return "", "", "", err
	}

	area := strings.TrimSpace(doc.Find("#tituloDetalleAviso h1").Text())
	// Clean articleType by replacing non-breaking spaces and trimming
	articleType := strings.TrimSpace(strings.ReplaceAll(doc.Find(".puntero.first-section").Text(), "\u00a0", " "))
	content := strings.TrimSpace(doc.Find("#cuerpoDetalleAviso").Text())

	if area == "" || articleType == "" || content == "" {
		return "", "", "", fmt.Errorf("failed to scrape required fields from %s", url)
	}

	return articleType, area, content, nil
}
